### PR TITLE
Add hierarchical ImageNet-like dataset format

### PR DESCRIPTION
**Pull request by itrushkin**
### Summary
CVS-143460

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: src/datumaro/components/importer.py [hunk 1]
```diff
@@ -197,7 +196,7 @@ def _change_context_root_path(context: FormatDetectionContext, path: str):
                 )
 
             for sub_dir in os.listdir(path):
-                if sub_dir.lower() in SUBSET_NAME_BLACKLIST:
+                if sub_dir.lower() not in ("train", "test", "val"):
```
Comment (vinnamkim) [inline src/datumaro/components/importer.py]: Why this change is needed?

File: src/datumaro/plugins/data_formats/imagenet.py
Comment (vinnamkim) [inline src/datumaro/plugins/data_formats/imagenet.py:1]: It is not desired to change the existing interface for consistency between versions. Is it impossible to add this new feature in additive manner?

**Review by vinnamkim — COMMENTED** [review]
Please make Windows CI green

File: src/datumaro/plugins/data_formats/imagenet.py
Comment (itrushkin) [reply src/datumaro/plugins/data_formats/imagenet.py:1]: The current `image_dir` importer implementation detects images at any subfolder level, leading to ambiguity with the newly added format. To resolve this, we have two options: either change the requirements for one of the formats (I did this) or introduce a new detection confidence level between `EXTREME_LOW` for `image_dir` and `LOW` for `imagenet`.

File: src/datumaro/components/importer.py (hunk 1 above)
Comment (itrushkin) [reply src/datumaro/components/importer.py]: This is necessary to distinguish between two importers with unlimited hierarchy levels: `nested_imagenet` and `nested_imagenet_with_subset_dirs`. Without this distinction, we cannot determine whether a folder is a subset directory or a top-level class.

File: src/datumaro/plugins/data_formats/imagenet.py
Comment (vinnamkim) [reply src/datumaro/plugins/data_formats/imagenet.py:1]: Yea sorry I misunderstood that `format=imagenet` is removed and there remains `format=nested_imagenet` only.

File: src/datumaro/components/importer.py (hunk 1 above)
Comment (vinnamkim) [inline src/datumaro/components/importer.py:199]: I meant what is the reason that we should remove `SUBSET_NAME_BLACKLIST`. Is it impossible such as

```suggestion
from datumaro.util.definitions import SUBSET_NAME_BLACKLIST, SUBSET_NAME_WHITELIST
# SUBSET_NAME_WHITELIST = ("train", "test", "val")

                if sub_dir.lower() in SUBSET_NAME_BLACKLIST or sub_dir.lower() not in SUBSET_NAME_WHITELIST:
```

p.s. I saw `("train", "test", "val")` is used other place so it seems possible to share it by declaring it at `datumaro.util.definitions`

Comment (vinnamkim) [reply src/datumaro/components/importer.py]: I meant what is the reason that we should remove `SUBSET_NAME_BLACKLIST`. Is it impossible such as

```suggestion
from datumaro.util.definitions import SUBSET_NAME_BLACKLIST, SUBSET_NAME_WHITELIST
# SUBSET_NAME_WHITELIST = ("train", "test", "val")

                if sub_dir.lower() in SUBSET_NAME_BLACKLIST or sub_dir.lower() not in SUBSET_NAME_WHITELIST:
```

p.s. I saw `("train", "test", "val")` is used other place so it seems possible to share it by declaring it at `datumaro.util.definitions`

Comment (itrushkin) [reply src/datumaro/components/importer.py]: If the whitelist condition is not fulfilled, the `sub_dir` is considered to belong to the blacklist, and vice versa. If the directory name is not on either list, it is treated as if it belongs to the blacklist and is not processed. This logically equals to process only whitelist condition.

Comment (vinnamkim) [reply src/datumaro/components/importer.py]: Ah ok then can you move it to `datumaro.util.definitions` to be shared in other source files such as `src/datumaro/plugins/data_formats/imagenet.py`

**Review by vinnamkim — APPROVED** [review]
